File: app/src/main/java/com/kaltura/kflow/ui/MediaPageFragment.java
package com.kaltura.kflow.ui;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatButton;

import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.kaltura.client.enums.AssetReferenceType;
import com.kaltura.client.enums.AssetType;
import com.kaltura.client.enums.PinType;
import com.kaltura.client.enums.RuleType;
import com.kaltura.client.services.AssetService;
import com.kaltura.client.services.BookmarkService;
import com.kaltura.client.services.PinService;
import com.kaltura.client.services.ProductPriceService;
import com.kaltura.client.services.UserAssetRuleService;
import com.kaltura.client.types.Asset;
import com.kaltura.client.types.BookmarkFilter;
import com.kaltura.client.types.ListResponse;
import com.kaltura.client.types.ProductPriceFilter;
import com.kaltura.client.types.UserAssetRule;
import com.kaltura.client.types.UserAssetRuleFilter;
import com.kaltura.client.utils.request.MultiRequestBuilder;
import com.kaltura.client.utils.request.RequestBuilder;
import com.kaltura.kflow.R;
import com.kaltura.kflow.manager.PhoenixApiManager;
import com.kaltura.kflow.ui.debug.DebugFragment;
import com.kaltura.kflow.ui.main.MainActivity;
import com.kaltura.kflow.ui.player.PlayerFragment;
import com.kaltura.kflow.utils.Utils;

import java.util.List;

/**
 * Created by alex_lytvynenko on 27.11.2018.
 */
public class MediaPageFragment extends DebugFragment implements View.OnClickListener {

    private TextInputEditText mMediaId;
    private TextInputEditText mPin;
    private AppCompatButton mPlay;
    private AppCompatButton mGetProductPrice;
    private AppCompatButton mGetBookmark;
    private AppCompatButton mGetAssetRules;
    private AppCompatButton mCheckAll;
    private AppCompatButton mInsertPin;
    private LinearLayout mPinLayout;
    private TextInputLayout mPinInputLayout;
    private Asset mAsset;
    private int mParentalRuleId;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_media_page, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ((MainActivity) requireActivity()).getSupportActionBar().setTitle("Media page");

        mMediaId = getView().findViewById(R.id.id);
        mPlay = getView().findViewById(R.id.play_asset);
        mGetProductPrice = getView().findViewById(R.id.get_product_price);
        mGetBookmark = getView().findViewById(R.id.get_bookmark);
        mGetAssetRules = getView().findViewById(R.id.get_asset_rules);
        mCheckAll = getView().findViewById(R.id.check_all);
        mInsertPin = getView().findViewById(R.id.insert_pin);
        mPinLayout = getView().findViewById(R.id.pin_layout);
        mPin = getView().findViewById(R.id.pin);
        mPinInputLayout = getView().findViewById(R.id.pin_input_layout);

        mPlay.setOnClickListener(this);
        mGetProductPrice.setOnClickListener(this);
        mGetBookmark.setOnClickListener(this);
        mGetAssetRules.setOnClickListener(this);
        mCheckAll.setOnClickListener(this);
        mInsertPin.setOnClickListener(this);
        getView().findViewById(R.id.get).setOnClickListener(this);

        validateButtons();
    }

    @Override
    public void onClick(View view) {
        Utils.hideKeyboard(getView());

        switch (view.getId()) {
            case R.id.play_asset: {
                playAsset();
                break;
            }
            case R.id.get: {
                getAssetRequest(mMediaId.getText().toString());
                break;
            }
            case R.id.get_product_price: {
                getProductPriceRequest(mMediaId.getText().toString());
                break;
            }
            case R.id.get_bookmark: {
                getBookmarkRequest(mMediaId.getText().toString());
                break;
            }
            case R.id.get_asset_rules: {
                getAssetRulesRequest(mMediaId.getText().toString());
                break;
            }
            case R.id.check_all: {
                checkAllTogetherRequest(mMediaId.getText().toString());
                break;
            }
            case R.id.insert_pin: {
                if (mPinInputLayout.getVisibility() == View.GONE) {
                    showPinInput();
                } else {
                    checkPinRequest(mPin.getText().toString());
                }
                break;
            }
        }
    }

    private void getAssetRequest(String assetId) {
        if (Utils.hasInternetConnection(requireContext())) {

            mAsset = null;
            mParentalRuleId = 0;
            mPin.setText("");

            validateButtons();

            RequestBuilder requestBuilder = AssetService.get(assetId, AssetReferenceType.MEDIA)
                    .setCompletion(result -> {
                        if (result.isSuccess()) {
                            if (result.results != null) mAsset = result.results;
                            validateButtons();
                        }
                    });
            PhoenixApiManager.execute(requestBuilder);
            clearDebugView();
        } else {
            Toast.makeText(requireContext(), "No Internet connection", Toast.LENGTH_SHORT).show();
        }
    }

    private void getProductPriceRequest(String assetId) {
        if (Utils.hasInternetConnection(requireContext())) {

            ProductPriceFilter productPriceFilter = new ProductPriceFilter();
            productPriceFilter.setFileIdIn(assetId);

            RequestBuilder requestBuilder = ProductPriceService.list(productPriceFilter);
            PhoenixApiManager.execute(requestBuilder);
            clearDebugView();
        } else {
            Toast.makeText(requireContext(), "No Internet connection", Toast.LENGTH_SHORT).show();
        }
    }

    private void getBookmarkRequest(String assetId) {
        if (Utils.hasInternetConnection(requireContext())) {

            BookmarkFilter bookmarkFilter = new BookmarkFilter();
            bookmarkFilter.setAssetIdIn(assetId);
            bookmarkFilter.setAssetTypeEqual(AssetType.MEDIA);

            RequestBuilder requestBuilder = BookmarkService.list(bookmarkFilter);
            PhoenixApiManager.execute(requestBuilder);
            clearDebugView();
        } else {
            Toast.makeText(requireContext(), "No Internet connection", Toast.LENGTH_SHORT).show();
        }
    }

    private void getAssetRulesRequest(String assetId) {
        if (Utils.hasInternetConnection(requireContext())) {

            if (TextUtils.isDigitsOnly(assetId)) {
                UserAssetRuleFilter userAssetRuleFilter = new UserAssetRuleFilter();
                userAssetRuleFilter.setAssetTypeEqual(1);
                userAssetRuleFilter.setAssetIdEqual(Long.parseLong(assetId));

                RequestBuilder requestBuilder = UserAssetRuleService.list(userAssetRuleFilter).setCompletion(result -> {
                    if (result.isSuccess()) {
                        if (result.results != null && result.results.getObjects() != null) {
                            handleUserRules(result.results.getObjects());
                        }
                        validateButtons();
                    }
                });
                PhoenixApiManager.execute(requestBuilder);
                clearDebugView();
            } else {
                Toast.makeText(requireContext(), "Wrong input", Toast.LENGTH_SHORT).show();
            }
        } else {
            Toast.makeText(requireContext(), "No Internet connection", Toast.LENGTH_SHORT).show();
        }
    }

    private void checkAllTogetherRequest(String assetId) {
        if (Utils.hasInternetConnection(requireContext())) {

            if (TextUtils.isDigitsOnly(assetId)) {
                MultiRequestBuilder multiRequestBuilder = new MultiRequestBuilder();

                // product price request
                ProductPriceFilter productPriceFilter = new ProductPriceFilter();
                productPriceFilter.setFileIdIn(assetId);
                multiRequestBuilder.add(ProductPriceService.list(productPriceFilter));

                // bookmark request
                BookmarkFilter bookmarkFilter = new BookmarkFilter();
                bookmarkFilter.setAssetIdIn(assetId);
                bookmarkFilter.setAssetTypeEqual(AssetType.MEDIA);
                multiRequestBuilder.add(BookmarkService.list(bookmarkFilter));

                // asset rules request
                UserAssetRuleFilter userAssetRuleFilter = new UserAssetRuleFilter();
                userAssetRuleFilter.setAssetTypeEqual(1);
                userAssetRuleFilter.setAssetIdEqual(Long.parseLong(assetId));
                multiRequestBuilder.add(UserAssetRuleService.list(userAssetRuleFilter));

                multiRequestBuilder.setCompletion(result -> {
                    if (result.isSuccess()) {
                        if (result.results != null && result.results.get(2) != null) {
                            handleUserRules(((ListResponse) result.results.get(2)).getObjects());
                        }
                        validateButtons();
                    }
                });

                PhoenixApiManager.execute(multiRequestBuilder);
                clearDebugView();
            } else {
                Toast.makeText(requireContext(), "Wrong input", Toast.LENGTH_SHORT).show();
            }
        } else {
            Toast.makeText(requireContext(), "No Internet connection", Toast.LENGTH_SHORT).show();
        }
    }

    private void checkPinRequest(String pin) {
        if (Utils.hasInternetConnection(requireContext())) {

            if (TextUtils.isDigitsOnly(pin)) {
                RequestBuilder requestBuilder = PinService.validate(pin, PinType.PARENTAL, mParentalRuleId);
                PhoenixApiManager.execute(requestBuilder);
                clearDebugView();
            } else {
                Toast.makeText(requireContext(), "Wrong input", Toast.LENGTH_SHORT).show();
            }
        } else {
            Toast.makeText(requireContext(), "No Internet connection", Toast.LENGTH_SHORT).show();
        }
    }

    private void showPinInput() {
        mPinInputLayout.setVisibility(View.VISIBLE);
        mInsertPin.setText("Check pin");
        Utils.showKeyboard(mPin);
    }

    private void playAsset() {
        PlayerFragment assetListFragment = PlayerFragment.newInstance(mAsset);
        requireActivity().getSupportFragmentManager().beginTransaction()
                .replace(R.id.container, assetListFragment)
                .addToBackStack(null)
                .commit();
    }

    private void validateButtons() {
        int visibility = mAsset == null ? View.GONE : View.VISIBLE;
        mPlay.setVisibility(visibility);
        mGetProductPrice.setVisibility(visibility);
        mGetBookmark.setVisibility(visibility);
        mGetAssetRules.setVisibility(visibility);
        mCheckAll.setVisibility(visibility);
        validatePinLayout();
    }

    private void validatePinLayout() {
        if (mParentalRuleId > 0) {
            mPinLayout.setVisibility(View.VISIBLE);
        } else {
            mPin.setText("");
            mPinLayout.setVisibility(View.GONE);
            mPinInputLayout.setVisibility(View.GONE);
            mInsertPin.setText("Insert pin");
        }
    }

    private void handleUserRules(List<UserAssetRule> userAssetRules) {
        for (UserAssetRule userAssetRule : userAssetRules) {
            if (userAssetRule.getRuleType() == RuleType.PARENTAL) {
                mParentalRuleId = userAssetRule.getId().intValue();
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Utils.hideKeyboard(getView());
        PhoenixApiManager.cancelAll();
    }

    @Override
    protected int getDebugViewId() {
        return R.id.debug_view;
    }
}

File: app/src/main/java/com/kaltura/kflow/presentation/mediaPage/MediaPageFragment.kt
package com.kaltura.kflow.presentation.mediaPage

import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isGone
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.navArgs
import com.github.leandroborgesferreira.loadingbutton.presentation.State
import com.kaltura.client.enums.RuleType
import com.kaltura.client.types.*
import com.kaltura.kflow.R
import com.kaltura.kflow.databinding.FragmentFavoritesBinding
import com.kaltura.kflow.databinding.FragmentMainBinding
import com.kaltura.kflow.databinding.FragmentMediaPageBinding
import com.kaltura.kflow.presentation.base.SharedTransitionFragment
import com.kaltura.kflow.presentation.debug.DebugView
import com.kaltura.kflow.presentation.extension.*
import com.kaltura.kflow.presentation.main.Feature
import org.koin.androidx.viewmodel.ext.android.viewModel

/**
 * Created by alex_lytvynenko on 27.11.2018.
 */
class MediaPageFragment : SharedTransitionFragment(R.layout.fragment_media_page) {

    private val viewModel: MediaPageViewModel by viewModel()
    private val args: MediaPageFragmentArgs by navArgs()
    private var parentalRuleId = 0
    private var asset: Asset? = null

    override val feature by lazy {
        when {
            args.isKeepAlive -> Feature.KEEP_ALIVE
            else -> Feature.MEDIA_PAGE
        }
    }
    private var _binding: FragmentMediaPageBinding? = null
    private val binding get() = _binding!!
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        _binding = FragmentMediaPageBinding.inflate(inflater,container,false)
        val view = binding.root
        return view
    }
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.playAsset.setOnClickListener {
            navigate(MediaPageFragmentDirections.navigateToPlayer(args.isKeepAlive, asset = asset!!))
        }
        binding.getProductPrice.setOnClickListener {
            hideKeyboard()
            getProductPriceRequest(binding.mediaId.string)
        }
        binding.getBookmark.setOnClickListener {
            hideKeyboard()
            getBookmarkRequest(binding.mediaId.string)
        }
        binding.getAssetRules.setOnClickListener {
            hideKeyboard()
            getAssetRulesRequest(binding.mediaId.string)
        }
        binding.checkAll.setOnClickListener {
            hideKeyboard()
            checkAllTogetherRequest(binding.mediaId.string)
        }
        binding.insertPin.setOnClickListener {
            hideKeyboard()
            if (binding.pinInputLayout.isGone) {
                showPinInput()
            } else {
                checkPinRequest(binding.pin.string)
            }
        }
        binding.get.setOnClickListener {
            hideKeyboard()
            getAssetRequest(binding.mediaId.string)
        }
        validateButtons()
    }

    override fun subscribeUI() {
        observeResource(viewModel.asset,
                error = { binding.get.error(lifecycleScope) },
                success = {
                    binding.get.success(lifecycleScope)
                    asset = it
                    validateButtons()
                })
        observeResource(viewModel.productPrices,
                error = { binding.getProductPrice.error(lifecycleScope) },
                success = { binding.getProductPrice.success(lifecycleScope) })
        observeResource(viewModel.bookmarks,
                error = { binding.getBookmark.error(lifecycleScope) },
                success = { binding.getBookmark.success(lifecycleScope) })
        observeResource(viewModel.userAssetRules,
                error = {
                    if (binding.getAssetRules.getState() == State.PROGRESS) binding.getAssetRules.error(lifecycleScope)
                    else binding.checkAll.error(lifecycleScope)
                },
                success = {
                    if (binding.getAssetRules.getState() == State.PROGRESS) binding.getAssetRules.success(lifecycleScope)
                    else binding.checkAll.success(lifecycleScope)
                    handleUserRules(it)
                    validateButtons()
                })
    }

    private fun getAssetRequest(assetId: String) {
        withInternetConnection {
            clearDebugView()
            clearInputLayouts()

            if (assetId.isEmpty()) {
                binding.mediaIdInputLayout.showError("Empty media ID")
                return@withInternetConnection
            }

            asset = null
            parentalRuleId = 0
            binding.pin.string = ""
            validateButtons()
            binding.get.startAnimation {
                viewModel.getAsset(assetId)
            }
        }
    }

    private fun getProductPriceRequest(assetId: String) {
        withInternetConnection {
            clearDebugView()
            clearInputLayouts()

            if (assetId.isEmpty()) {
                binding.mediaIdInputLayout.showError("Empty media ID")
                return@withInternetConnection
            }

            binding.getProductPrice.startAnimation {
                viewModel.getProductPrice(assetId)
            }
        }
    }

    private fun getBookmarkRequest(assetId: String) {
        withInternetConnection {
            clearDebugView()
            clearInputLayouts()

            if (assetId.isEmpty()) {
                binding.mediaIdInputLayout.showError("Empty media ID")
                return@withInternetConnection
            }

            binding.getBookmark.startAnimation {
                viewModel.getBookmark(assetId)
            }
        }
    }

    private fun getAssetRulesRequest(assetId: String) {
        withInternetConnection {
            clearDebugView()
            clearInputLayouts()

            if (assetId.isEmpty()) {
                binding.mediaIdInputLayout.showError("Empty media ID")
                return@withInternetConnection
            }
            if (TextUtils.isDigitsOnly(assetId).not()) {
                binding.mediaIdInputLayout.showError("Wrong input")
                return@withInternetConnection
            }

            binding.getAssetRules.startAnimation {
                viewModel.getAssetRules(assetId)
            }
        }
    }

    private fun checkAllTogetherRequest(assetId: String) {
        withInternetConnection {
            clearDebugView()
            clearInputLayouts()

            if (assetId.isEmpty()) {
                binding.mediaIdInputLayout.showError("Empty media ID")
                return@withInternetConnection
            }
            if (TextUtils.isDigitsOnly(assetId).not()) {
                binding.mediaIdInputLayout.showError("Wrong input")
                return@withInternetConnection
            }

            binding.checkAll.startAnimation {
                viewModel.checkAllTogether(assetId)
            }
        }
    }

    private fun checkPinRequest(pin: String) {
        withInternetConnection {
            if (TextUtils.isDigitsOnly(pin)) {
                clearDebugView()
                viewModel.checkPin(pin, parentalRuleId)
            } else {
                toast("Wrong input")
            }
        }
    }

    private fun clearInputLayouts() {
        binding.mediaIdInputLayout.hideError()
    }

    private fun showPinInput() {
        binding.pinInputLayout.visible()
        binding.insertPin.text = "Check pin"
        showKeyboard(binding.pin)
    }

    private fun validateButtons() {
        val isVisible = asset != null
        binding.playAsset.visibleOrGone(isVisible)
        binding.getProductPrice.visibleOrGone(isVisible)
        binding.getBookmark.visibleOrGone(isVisible)
        binding.getAssetRules.visibleOrGone(isVisible)
        binding.checkAll.visibleOrGone(isVisible)
        validatePinLayout()
    }

    private fun validatePinLayout() {
        if (parentalRuleId > 0) {
            binding.pinLayout.visible()
        } else {
            binding.pin.string = ""
            binding.pinLayout.gone()
            binding.pinInputLayout.gone()
            binding.insertPin.text = "Insert pin"
        }
    }

    private fun handleUserRules(userAssetRules: List<UserAssetRule>) {
        userAssetRules.forEach {
            if (it.ruleType == RuleType.PARENTAL) {
                parentalRuleId = it.id.toInt()
            }
        }
    }
}